### PR TITLE
capybara

### PR DESCRIPTION
**Pull request by Sova171**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -59,6 +59,8 @@ gem 'tzinfo-data', platforms: %i[mingw mswin x64_mingw jruby]
 
 gem 'bootsnap', require: false
 
+gem 'haml-rails'
```
Comment (anastasiabalaban) [inline Gemfile:62]: Again, why did you add this gem? You've already added hamlit gem

File: app/views/books/index.html.haml [hunk 2]
```diff
@@ -1,20 +1,22 @@
 %h1.card-title= t('.books')
-
-.dropdown
-  %button#dropdownMenuButton1.btn.dropdown-toggle{"aria-expanded" => "false", "data-bs-toggle" => "dropdown", type: "button"}
-    = t('.categories')
-  %ul.dropdown-menu{ "aria-labelledby" => "dropdownMenuButton1" }
-    %li
-      = link_to t('.all'), root_path, class: "dropdown-item"
-    - @facade.categories.each do |category|
+- if @facade.books.blank?
```
Comment (anastasiabalaban) [inline app/views/books/index.html.haml:2]: add a method to the facade where you will check if there are any books.

Also, please check differences between blank?, empty?, any?, present?, exists? methods.

File: db/seeds/books.rb [hunk 3]
```diff
@@ -8,7 +8,8 @@ def create_book
     publisher:    Faker::Sports::Basketball.team,
     description:  Faker::TvShows::FamilyGuy.quote,
     pages_count:  rand(10..320),
-    published_at: Faker::TvShows::Friends.quote
+    published_at: Faker::TvShows::Friends.quote,
+    category:     %w[horror biography detective fantasy mystery other].sample
```
Comment (anastasiabalaban) [inline db/seeds/books.rb:12]: What if categories list changes in the future? Could you use `Book.categories.sample`?

File: spec/factories/information.rb [hunk 4]
```diff
@@ -9,6 +9,7 @@
   factory :book do
     title { Faker::JapaneseMedia::OnePiece.location }
     description { Faker::TvShows::FamilyGuy.quote }
+    category { %w[horror biography detective fantasy mystery other].sample }
```
Comment (anastasiabalaban) [inline spec/factories/information.rb:12]: same here

File: spec/features/book_index_redirect_spec.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  before(:each) do
+    create(:book)
+    Book.reindex
+
+    visit root_path
+  end
+
+  context 'user search:' do
+    it 'specific world' do
+      fill_in 'Book title', with: 'Yard'
```
Comment (anastasiabalaban) [inline spec/features/book_index_redirect_spec.rb:15]: add only your test expectations to `it` statement and move everything else to before block or let method helper

File: spec/features/book_index_redirect_spec.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+      click_button 'Search'
+      expect(page).to have_current_path('/searches?search=Yard')
+    end
+
+    it 'nothing' do
+      fill_in 'Book title', with: ''
```
Comment (anastasiabalaban) [inline spec/features/book_index_redirect_spec.rb:21]: and don't repeat yourself. For example, you can use `let(:title) { book.title }` and then you can move part of your test to the before block:
```
before do
  fill_in 'Book title', with: title
  click_button 'Search'
end
```

File: spec/features/book_index_redirect_spec.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+      click_button 'Search'
+      expect(page).to have_current_path(root_path)
+    end
+  end
+
+  context 'when user pick category:' do
```
Comment (anastasiabalaban) [inline spec/features/book_index_redirect_spec.rb:27]: user pick**s**

File: spec/features/book_index_redirect_spec.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    it 'all categories' do
```
Comment (anastasiabalaban) [inline spec/features/book_index_redirect_spec.rb:28]: `it 'returns all books' do` -- don't forget about verbs' existence. Also, the main reason for your test is to test BOOKS and you should check if correct books exists in the page

File: spec/features/book_index_redirect_spec.rb [hunk 9]
```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.feature 'Book index redirect', type: :feature do
```
Comment (anastasiabalaban) [inline spec/features/book_index_redirect_spec.rb:5]: Strange separation 🤔
It is enough to have just `book_index_spec.rb` where you will test books and their search

File: spec/features/book_index_spec.rb [hunk 10]
```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.feature 'Book index', type: :feature do
+  context 'when there are no books on the' do
+    let(:nothing) { 'Nothing here yet' }
+
+    before(:each) do |test|
+      create(:book, title: 'JO', category: 'other')          unless test.metadata[:focus]
```
Comment (anastasiabalaban) [inline spec/features/book_index_spec.rb:10]: what does `test.metadata` mean? why do you need this?

File: spec/features/book_index_spec.rb [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    end
+
+    it 'first page' do
+      visit root_path
+      expect(page).to have_content('JO')
+      expect(page).to_not have_content('Giovanna')
+    end
+
+    it 'second page' do
+      visit '/?page=2'
```
Comment (anastasiabalaban) [inline spec/features/book_index_spec.rb:51]: There is no need to test pagination. You are using "pagy" gem and let's assume it works fine. I guess it will be fine if you will create max 4 books for your spec

**Review by anastasiabalaban — COMMENTED** [review]
Please review all my comments and refactor all your test examples 🙌

File: spec/features/book_index_spec.rb (hunk 10 above)
Comment (Sova171) [reply spec/features/book_index_spec.rb:10]: It's for before(:each) block except some specific tests

File: app/facades/books/index_facade.rb [hunk 12]
```diff
@@ -24,6 +24,10 @@ def categories
       Book.categories
     end
 
+    def book_presence
```
Comment (anastasiabalaban) [inline app/facades/books/index_facade.rb:27]: this method returns true or false, so it is better to name it with a question mark at the end. For example, `books_present?`

File: spec/features/book_index_spec.rb [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+    context 'root page' do
+      it 'show empty message' do
+        visit root_path
+        expect(page).to have_content(nothing)
+      end
+    end
+
+    context 'page' do
+      before(:each) do
```
Comment (anastasiabalaban) [inline spec/features/book_index_spec.rb:17]: just `before do`. I guess this should work

File: spec/features/book_index_spec.rb [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    let(:nothing) { 'Nothing here yet' }
+
+    context 'root page' do
+      it 'show empty message' do
+        visit root_path
+        expect(page).to have_content(nothing)
+      end
+    end
+
+    context 'page' do
```
[the remark this answers is not in the document]
Comment (anastasiabalaban) [reply spec/features/book_index_spec.rb:16]: and there should be two separate contexts:
- when books exist
- when books do not exist
(not the one related to books inside "no books" context)

File: spec/features/book_index_spec.rb [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+        end
+
+        it 'show no results message' do
+          expect(page).to have_content('No results')
+        end
+      end
+    end
+  end
+
+  context 'when user is try to' do
```
Comment (anastasiabalaban) [inline spec/features/book_index_spec.rb:49]: 'when user tries to'

File: spec/features/book_index_spec.rb [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+      create(:book, title: 'Higashikata', category: 'horror')
+      create(:book, title: 'Giovanna', category: 'detective')
+      Book.reindex
+
+      visit root_path
+    end
+
+    context 'search for an existing book' do
+      before do
+        fill_in 'Book title', with: 'Giovanna'
```
Comment (anastasiabalaban) [inline spec/features/book_index_spec.rb:62]: you can create list of random books (create_list), choose one book (books.sample) and use
`fill_in 'Book title', with: book.title` and then use `book.title` in spec

File: spec/features/book_index_spec.rb [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+
+      it 'return correct page' do
+        expect(page).to have_current_path('/searches?search=Giovanna')
+      end
+    end
+
+    context 'picks category: horror' do
+      before do
+        find('#filter-categories').click
+        find('.dropdown-item', text: 'horror').click
```
Comment (anastasiabalaban) [inline spec/features/book_index_spec.rb:78]: for categories you can separately create book and use it in tests (horror_book.title): 
`let(:horror_book) { create(:book,  category: :horror }`

File: spec/features/user_registration_spec.rb [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+        fill_in 'Password confirmation', with: 'Password'
+      end
+
+      context 'where are all correct' do
+        before do
+          fill_in 'Email', with: 'vlad@gmail.com'
+        end
+
+        it 'will create a user account and redirect to root path' do
+          expect { subject }.to change { User.count }.by(1)
```
Comment (anastasiabalaban) [inline spec/features/user_registration_spec.rb:36]: user count checks your unit service specs, feature specs tests stuff on the page which cannot be tested in unit specs

File: spec/features/user_registration_spec.rb [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+          expect(page).to have_current_path(root_path)
+        end
+      end
+
+      context 'where is invalid email' do
```
Comment (anastasiabalaban) [inline spec/features/user_registration_spec.rb:41]: when email is invalid

File: spec/features/book_index_spec.rb [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+      end
+
+      it 'show no results message' do
+        expect(page).to have_content('No results')
+      end
+    end
+  end
+
+  context 'when book exist and user tries to' do
+    let(:list) { Book.all }
```
Comment (anastasiabalaban) [inline spec/features/book_index_spec.rb:50]:
```
let(:book_list)   { create_list(:book, 4) }
let(:random_book) { book_list.sample }
```

File: spec/features/user_registration_spec.rb [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+      end
+    end
+
+    context 'with fields' do
+      before do
+        fill_in 'Password',              with: 'Password'
+        fill_in 'Password confirmation', with: 'Password'
+      end
+
+      context 'where are all correct' do
```
Comment (anastasiabalaban) [inline spec/features/user_registration_spec.rb:30]: 'when params are valid' or 'when fields are valid'

File: spec/features/user_registration_spec.rb [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+require 'rails_helper'
+
+RSpec.feature 'User Registration', type: :feature do
+  subject { click_button 'Sign up' }
+
+  before do
+    visit new_user_registration_path
+  end
+
+  context 'when user tries submits to the form' do
```
Comment (anastasiabalaban) [inline spec/features/user_registration_spec.rb:12]: Remove this context. It s clear that user tries to submit the form because it is User Registration spec.
